### ISSUE TITLE
Ceph version management with Rook version

### ISSUE THREAD

**Issue by leseb** [rook/rook#2632]
Since I've started my work on implementing msgr2 support in Rook (https://github.com/rook/rook/pull/2559), I add to implement a Ceph version check, to perform certain operations whether we run Nautilus and above or earlier versions.
Currently Rook support Luminous, Mimic and Nautilus (let's say partially)

The exceptions are going to grow over time, that's a given. The more they grow the more complex the Rook code will be to maintain and years from now we are going to regret it. At first we think, "yes that's not so much" and then we keep addind more and more exceptions. I realize this might break Rook's promise to have a single operator *to rule them all* but in the long run that's not worth it.

To mitigate that, I believe we should start pinning the support for N number of version of Ceph for a certain version of the Rook operator. One major version of the operator could support up to two Ceph version (so we can handle upgrades).

Rook 1.0 is around the corner and I believe it's a good opportunity to make it Ceph Nautilus only. So basically we would have something like:

* Rook 0.X supports Luminous and Mimic
* Rook 1.0 supports Nautilus (upgrade path included from Mimic) and Octopus
* Rook 1.1 supports  P (upgrade path included from Octopus) and Q

So to go from Mimic to Nautilus, you will have to upgrade the operator first, then trigger your Ceph upgrade to Nautilus.

@BlaineEXE @travisn @jbw976 @galexrt, @liewegas, @rootfs  please share your thoughts :).

**Comment by travisn** [comment]
Great discussion to start. My initial thoughts on this would be:
- The latest two releases of Ceph are supported. 
- Example yamls would use the latest supported release
- We need a transition release when we stop supporting a version. For example, this means in v1.0 we stop supporting Luminous, but since users upgrade Rook before updating Ceph, we still need supporting code in 1.0 for Luminous.

This would translate to the following versions supported:
- v0.9: Luminous and Mimic. Nautilus is experimental.
- v1.0: Mimic and Nautilus. Octopus is experimental. 
  - Users are expected to migrate off Luminous after updating to v1.0, so technically Luminous is still supported for this transition release.
- v1.1: Mimic and Nautilus. Octopus is experimental. Code can be removed for supporting Luminous

**Comment by Bessonov** [comment]
> I realize this might break Rook's promise to have a single operator _to rule them all_ [...]

Just an offtopic comment from a beginner rook user: please, don't support everything. Please support less, but well. I've experienced multiple troubles and instability of rook on my test cluster and sometime I just throw away and setup rook again, because nothing worked. I definitely don't welcome support of CockroachDB (although I'm on the way to migrate to it), cassandra and others. I really wish just a rock stable ceph support. I don't need an operator for others databases. It splits the community.

On supported releases. From my personal view there is no need to support experimental releases. Just an suggestion:
* v0.9: Luminous and Mimic
* v1.0-alpha: Mimic and experimental Nautilus.
* v1.0: Mimic and Nautilus.
* v1.1-alpha: Nautilus and experimental P.
* v1.1: Nautilus and P.

Less work for you, more done for us :+1: 

May be another suggestion: Use major version to indicate target ceph release. For example 1.x.y Nautilus, 2.x.y Octopus, 3.x.y P and so on.

**Comment by travisn** [comment]
@Bessonov Did you find if there are issues tracking the problems you hit? We are certainly very interested in fixing the issues that have caused you to restart the cluster, whether it is better documentation for new users, or other fixes.

From my view, we wouldn't provide official support for experimental releases. It is really only allowed for the development community to test the upcoming version of Ceph so that we can fully support it soon after it's released.

**Comment by Bessonov** [comment]
@travisn I see multiple strange things, but it can be a miss-configuration from my side. Just ignore it for now. Now I have much less movement in my cluster and will take care of such things :+1:

**Comment by allen13** [comment]
Here's quick example of a multi-version support bug caused by the complexities of detecting versions via a kubernetes job:

2019-02-11 20:38:53.679468 I \| op-k8sutil: Removing previous job rook-ceph-detect-version to start a new one
--
  | 2019-02-11 20:38:53.691917 I \| op-k8sutil: batch job rook-ceph-detect-version still exists

Fix:
kubectl delete jobs --all -n rook-ceph
job.batch "rook-ceph-detect-version" deleted
job.batch "rook-ceph-osd-prepare-abc1" deleted

Basically the job deletion doesn't always work. If the operator pod gets restarted during this state it never makes it to the provisioner code and pvc create/delete stops working.

**Comment by mmack** [comment]
> > I realize this might break Rook's promise to have a single operator _to rule them all_ [...]
> 
> Just an offtopic comment from a beginner rook user: please, don't support everything. Please support less, but well. I've experienced multiple troubles and instability of rook on my test cluster and sometime I just throw away and setup rook again, because nothing worked. I definitely don't welcome support of CockroachDB (although I'm on the way to migrate to it), cassandra and others. I really wish just a rock stable ceph support. I don't need an operator for others databases. It splits the community.

Exact same thoughts here...  Ceph is a beast at it's own, just look at the open issues: 5 cockroach, 7 cassandra and 212 ceph.

**Comment by BlaineEXE** [comment]
Seb has a good point, and I share his concerns. I think it seems feasible to support 2 versions of Ceph per release. I think it's important to note that Rook's release schedule differs quite a bit from Ceph's. Rook is released quarterly while Ceph is released on a 9-month cadence if I remember correctly.

**My thoughts on a possible support cadence:**
Rook v0.9 supports Luminous and Mimic
Rook v1.0 supports Luminous only for upgrade to Mimic, Mimic and Nautilus
Rook v1.1 supports Mimic and Nautilus
Rook v1.2 supports Mimic and Nautilus (Octopus support experimental); users are expected to migrate from Mimic to Nautilus
Rook v1.3 supports Nautilus and Octopus
Rook v1.4 supports Nautilus and Octopus
Rook v1.5 supports Nautilus and Octopus (P support experimental); users expected to migrate to Octopus
Rook v1.6 supports Octopus and P
Rook v1.7 supports Octopus and P
Rook v1.8 supports Octopus and P (Q support experimental); users are expected to migrate to P
Rook v1.9 supports P and Q
... continue on in the pattern

There is the option to do as Travis suggested and support the previous version (as in v1.0) only so that an upgrade can be done for Releases v1.3, v1.6, v1.9

-----

I also want to point out a logical fallacy I see: there is not a directly comparable relationship between the number of open issues for Ceph compared to other backends and the quality of Ceph code versus other backends. There are many other factors at play. Ceph is Rook's oldest backend by far, and it has had 3 very active developers for the past 3 releases (soon five). Considering these factors, Ceph should have more issues than others. Also, "issue" has more than one meaning on GitHub: some are feature requests and not bugs.

**Comment by conet** [comment]
If this is implemented I assume that it would mean that the operator would refuse to upgrade to a version of ceph which is not supported. I would avoid cases [like this](https://github.com/rook/rook/issues/2397#issuecomment-478161551) where one would end up with a partially working cluster without any immediate fix available:
- running the latest rook release `0.9.3`
- upgraded ceph from `ceph/ceph:v13.2.4-20190109` to `ceph/ceph:v14.2.0-20190319` (by patching the image in the CephCluster definition)
- stuck in a partial upgrade because the monitors were already upgraded to `ceph/ceph:v14.2.0-20190319` (I don't think ceph supports downgrade) and the manager fails to start because `0.9.3` is [lacking nautilus support](https://github.com/rook/rook/blob/master/pkg/daemon/ceph/client/mgr.go#L60), it's missing the `--force` when setting config

Is there any plan to release `0.9` with support for nautilus? I'm fine waiting for `1.0` but will the upgrade work in this state?

**Comment by travisn** [comment]
@conet Could you open an issue for this upgrade issue you hit? It's not expected that any upgrade would be performed by the operator if you use a nautilus image unless you set the `allowUnsupported` flag. With v0.9 releases nautilus isn't planned to be supported. v1.0 is planned to supported nautilus.

**Comment by conet** [comment]
@travisn I haven't set that flag so the question is what is the default behaviour, the doc states that:
> Currently only luminous and mimic are supported, so nautilus would require this to be set to true. Should be set to `false` in production.

There is no default value specified, if you think that an issues still needs to be opened let me know. I would be also curious about this: when `v1.0` is release will I be able to upgrade from this state? Currently the cluster is in `state: Updating` (still usable) and I see no other way to fix that then to upgrade to `v1.0` (when it will be available) but if the operator will only upgrade a stable cluster then I'm definitely stuck. Do you have other ideas to get out from this situation?

**Comment by conet** [comment]
I was able to recover the cluster after updating to `v1.0.0` and a few manual steps.
